apollo-router/src/query_planner/selection.rs

```diff
@@ -48,97 +48,171 @@ pub(crate) struct InlineFragment {
     let mut output = Object::with_capacity(selections.len());
     for selection in selections {
+        println!(
+            "execute_selection_set: selection {selection:?} from {}",
+            serde_json::to_string(&content).unwrap(),
+        );
+
         match selection {
-            Selection::Field(field) => {
-                if let Some((key, value)) = select_field(content, field, schema)? {
-                    if let Some(o) = output.get_mut(field.name.as_str()) {
-                        o.deep_merge(value);
-                    } else {
-                        output.insert(key.to_owned(), value);
+            Selection::Field(Field {
+                alias,
+                name,
+                selections,
+            }) => {
+                let selection_name = alias.as_deref().unwrap_or(name.as_str());
+                let field_type = current_type.and_then(|t| {
+                    schema.definitions.types.get(t).and_then(|ty| match ty {
+                        apollo_compiler::schema::ExtendedType::Object(o) => {
+                            o.fields.get(name.as_str()).map(|f| &f.ty)
+                        }
+                        apollo_compiler::schema::ExtendedType::Interface(i) => {
+                            i.fields.get(name.as_str()).map(|f| &f.ty)
+                        }
+                        _ => None,
+                    })
+                });
+
+                match content.get_key_value(selection_name) {
+                    None => {
+                        if name == "__typename" {
+                            // if the __typename field was missing but we can infer it, fill it
+                            if let Some(ty) = current_type {
+                                output.insert(
+                                    ByteString::from(selection_name.to_owned()),
+                                    Value::String(ByteString::from(ty.to_owned())),
+                                );
+                                continue;
+                            }
+                        }
+                        // the behaviour here does not align with the gateway: we should instead assume that
+                        // data is in the correct shape, and return a null (or even no value at all) on
+                        // missing fields. If a field was missing, it should have been nullified,
+                        // and if it was non nullable, the parent object would have been nullified.
+                        // Unfortunately, we don't validate subgraph responses yet
+                        if field_type
+                            .as_ref()
+                            .map(|ty| !ty.is_non_null())
+                            .unwrap_or(false)
+                        {
+                            output.insert(ByteString::from(selection_name.to_owned()), Value::Null);
+                        } else {
+                            return Value::Null;
+                        }
```

Comment:
I am not entirely confident about this part because it differs from the gateway's implementation. The rationale from the gateway's side is that a missing field in the data we select from should not have happened, and should have been caught by response validation, but I don't see where this validation is happening

Reply:
@trevor-scheer @clenfest is something you could help us figure out? 🙏

Reply:
From memory (a long time ago) I was under the impression that we never validated responses from subgraphs as this was an expensive operation.